### PR TITLE
Replace all FireCSharpEvents with BloomApi.post calls

### PR DESCRIPTION
**Pull request by StephenMcConnel**
This change is [Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289)

### CODE REVIEW

**Review by andrew-polk — CHANGES_REQUESTED** [review]
> Reviewed 9 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289)* status: all files reviewed, 2 unresolved discussions (waiting on @StephenMcConnel)

___
*[src/BloomExe/web/controllers/EditingViewApi.cs, line 90 at r1](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289#-MW11OITBLkItTarzDvX:-MW11OITBLkItTarzDvY:b-wlp9a5) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/25854cc3ed2748f2716a207962d2b9bb7476b50b/src/BloomExe/web/controllers/EditingViewApi.cs#L90)):*
> ```
> NotifyEditPagePainted?.Invoke();
> ```
I think this line can be simplified to 
`View.Model.NavigatingSoSuspendSaving = false;` (or perhaps call a method in model which does that?)
and the action removed. Or am I missing something?
___
*[src/BloomExe/web/controllers/EditingViewApi.cs, line 91 at r1](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289#-MW13Rhl6ChZacaU6eZC:-MW13Rhm8XBquAlS31X8:bps3zln) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/25854cc3ed2748f2716a207962d2b9bb7476b50b/src/BloomExe/web/controllers/EditingViewApi.cs#L91)):*
> ```
> NotifyEditPagePaintedFinished?.Invoke();
> ```
This (a static action defined in the API and implemented by a basically unrelated bit of code) doesn't *feel* right to me, but I haven't come up with a better implementation.
Maybe @johnthomson will have an idea.

**Review by JohnThomson — CHANGES_REQUESTED** [review]
Looks good. Few small points.

> Reviewed 9 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289#-:-MW1B_8gD0E5XEbeGiUM:bsuth62)* status: all files reviewed, 4 unresolved discussions (waiting on @andrew-polk and @StephenMcConnel)

___
*[src/BloomExe/web/controllers/EditingViewApi.cs, line 23 at r1](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289#-MW18rRm0VVjh92BNEaH:-MW18rRn3tiksZW53qys:bolts3x) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/25854cc3ed2748f2716a207962d2b9bb7476b50b/src/BloomExe/web/controllers/EditingViewApi.cs#L23)):*
> ```Smalltalk
> 			apiHandler.RegisterEndpointHandler("editView/getBookColors", HandleGetColors, true);
> 			apiHandler.RegisterEndpointHandler("editView/editPagePainted", HandleEditPagePainted, true);
> 			apiHandler.RegisterEndpointHandler("editView/saveToolboxSetting", HandleSaveToolboxSetting, false);
> ```

I'm nervous about this not being on the UI thread. We're calling a method of a Windows control, and the normal assumption is that such methods should be called on the UI thread. ToolboxView.SaveToolboxSettings() has a switch and calls at least three other methods; I haven't looked to see if they call yet others. If all of them are both thread-safe and safe to call on a non-UI thread, I think we should at least comment all of them to note that they have to be. But unless there's a reason not to just register this method as UI-thread (and if there is, please comment THAT), I'd be inclined to just switch it to UI-thread.
___
*[src/BloomExe/web/controllers/EditingViewApi.cs, line 91 at r1](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289#-MW13Rhl6ChZacaU6eZC:-MW17hfi3HArKXgQMIL7:b9kj8j3) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/25854cc3ed2748f2716a207962d2b9bb7476b50b/src/BloomExe/web/controllers/EditingViewApi.cs#L91)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> ```
> NotifyEditPagePaintedFinished?.Invoke();
> ```
This (a static action defined in the API and implemented by a basically unrelated bit of code) doesn't *feel* right to me, but I haven't come up with a better implementation.
Maybe @johnthomson will have an idea.
</blockquote></details>

Another approach I implemented at one point is to give EditingViewApi an event EditPagePainted. HandlePagePainted would then simply raise it; anything that cares can add and remove handlers. See 2e790371516167cbf77a2d5f8231adb995c168b3 ("Repair C# notification of page painted complete") on my fork.
___
*[src/BloomExe/web/controllers/EditingViewApi.cs, line 106 at r1](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289#-MW1AUm72jC35D2j0D1d:-MW1AUm72jC35D2j0D1e:b4kjuay) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/25854cc3ed2748f2716a207962d2b9bb7476b50b/src/BloomExe/web/controllers/EditingViewApi.cs#L106)):*
> ```Smalltalk
> 			var topicString = request.RequiredPostString();
> 			if (topicString == "<NONE>")
> 				topicString = "";	// required string cannot be empty
> ```

This is an odd comment on a line of code that makes it empty. Did you mean "can't be null"?  Oh, I see, you mean that the value passed to PostJson can't be empty, but that's an option we need here so "<NONE>" is used to mean it. Maybe on the previous line "a substitute for empty string, which is not allowed to be the result of RequiredPostJson"

**Review by andrew-polk — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289)* status: all files reviewed, 4 unresolved discussions (waiting on @JohnThomson and @StephenMcConnel)

___
*[src/BloomExe/web/controllers/EditingViewApi.cs, line 91 at r1](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289#-MW13Rhl6ChZacaU6eZC:-MW1Fvhs7ek0mlzqG8T5:bc6cfhz) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/25854cc3ed2748f2716a207962d2b9bb7476b50b/src/BloomExe/web/controllers/EditingViewApi.cs#L91)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Another approach I implemented at one point is to give EditingViewApi an event EditPagePainted. HandlePagePainted would then simply raise it; anything that cares can add and remove handlers. See 2e790371516167cbf77a2d5f8231adb995c168b3 ("Repair C# notification of page painted complete") on my fork.
</blockquote></details>

I thought of that, too, but couldn't convince myself that it was better at the time. You've convinced me, though.

Would it be better defined in the model (or perhaps view?...), though? And have the api simply tell the model we're done painting so the model can raise the event?

In other words, I think it would be better for WebThumbnailList to not be aware of EditingViewApi at all.

**Review by JohnThomson — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289)* status: all files reviewed, 4 unresolved discussions (waiting on @andrew-polk and @StephenMcConnel)

___
*[src/BloomExe/web/controllers/EditingViewApi.cs, line 91 at r1](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289#-MW13Rhl6ChZacaU6eZC:-MW1MMEECaes1OyDEseM:b37am5q) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/25854cc3ed2748f2716a207962d2b9bb7476b50b/src/BloomExe/web/controllers/EditingViewApi.cs#L91)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

I thought of that, too, but couldn't convince myself that it was better at the time. You've convinced me, though.

Would it be better defined in the model (or perhaps view?...), though? And have the api simply tell the model we're done painting so the model can raise the event?

In other words, I think it would be better for WebThumbnailList to not be aware of EditingViewApi at all.
</blockquote></details>

Another option is to use Autofac to decouple completely. Make a new subclass of Bloom.Event, like BookSavedEvent, and let AutoFac pass the one instance to whoever is interested. But defining an ordinary event on EditingModel is reasonable too. At this point there are only two clients. I agree, it's unfortunate that WebThumbnailList has to know about EditingViewApi, and it already knows about EditingModel. In some ways EditingView feels like an even more natural home for something so UI-related as an event raised on completing the painting of a page, but as yet WebThumbnailList doesn't need to know about EditingView, either.
If we can't find a place we're really satisfied this event belongs, that would argue for just letting Autofac own it.

**Review by JohnThomson — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289)* status: all files reviewed, 4 unresolved discussions (waiting on @andrew-polk and @StephenMcConnel)

___
*[src/BloomExe/web/controllers/EditingViewApi.cs, line 91 at r1](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289#-MW13Rhl6ChZacaU6eZC:-MW4xw0e88THIQXGb6KE:b-xbh9r6) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/25854cc3ed2748f2716a207962d2b9bb7476b50b/src/BloomExe/web/controllers/EditingViewApi.cs#L91)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Another option is to use Autofac to decouple completely. Make a new subclass of Bloom.Event, like BookSavedEvent, and let AutoFac pass the one instance to whoever is interested. But defining an ordinary event on EditingModel is reasonable too. At this point there are only two clients. I agree, it's unfortunate that WebThumbnailList has to know about EditingViewApi, and it already knows about EditingModel. In some ways EditingView feels like an even more natural home for something so UI-related as an event raised on completing the painting of a page, but as yet WebThumbnailList doesn't need to know about EditingView, either.
If we can't find a place we're really satisfied this event belongs, that would argue for just letting Autofac own it.
</blockquote></details>

Another reason to like the Autofac approach is that if we ever want to unit test anything related to this event, it's much easier to inject the SUT with an instance of a PagePaintFinished event than an instance of EditingView, EditingModel, or EditingViewApi

**Review by StephenMcConnel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289)* status: 5 of 9 files reviewed, 4 unresolved discussions (waiting on @andrew-polk and @JohnThomson)

___
*[src/BloomExe/web/controllers/EditingViewApi.cs, line 23 at r1](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289#-MW18rRm0VVjh92BNEaH:-MW1OPBoBoHK5H3f-McE:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/25854cc3ed2748f2716a207962d2b9bb7476b50b/src/BloomExe/web/controllers/EditingViewApi.cs#L23)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

I'm nervous about this not being on the UI thread. We're calling a method of a Windows control, and the normal assumption is that such methods should be called on the UI thread. ToolboxView.SaveToolboxSettings() has a switch and calls at least three other methods; I haven't looked to see if they call yet others. If all of them are both thread-safe and safe to call on a non-UI thread, I think we should at least comment all of them to note that they have to be. But unless there's a reason not to just register this method as UI-thread (and if there is, please comment THAT), I'd be inclined to just switch it to UI-thread.
</blockquote></details>

Done.
___
*[src/BloomExe/web/controllers/EditingViewApi.cs, line 90 at r1](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289#-MW11OITBLkItTarzDvX:-MW1OfsE380r5G85MCAU:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/25854cc3ed2748f2716a207962d2b9bb7476b50b/src/BloomExe/web/controllers/EditingViewApi.cs#L90)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> ```
> NotifyEditPagePainted?.Invoke();
> ```
I think this line can be simplified to 
`View.Model.NavigatingSoSuspendSaving = false;` (or perhaps call a method in model which does that?)
and the action removed. Or am I missing something?
</blockquote></details>

Done.
___
*[src/BloomExe/web/controllers/EditingViewApi.cs, line 91 at r1](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289#-MW13Rhl6ChZacaU6eZC:-MW52QJSCVPB5HiDe7Gn:b-60jwv) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/25854cc3ed2748f2716a207962d2b9bb7476b50b/src/BloomExe/web/controllers/EditingViewApi.cs#L91)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Another reason to like the Autofac approach is that if we ever want to unit test anything related to this event, it's much easier to inject the SUT with an instance of a PagePaintFinished event than an instance of EditingView, EditingModel, or EditingViewApi
</blockquote></details>

Done.  I changed to using an EventHandler in Editing Model.   (Autofac may be wonderful and perfect, but I don't understand how it works at all.)
___
*[src/BloomExe/web/controllers/EditingViewApi.cs, line 106 at r1](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289#-MW1AUm72jC35D2j0D1d:-MW1P6g-1ZgpO7-eT8-l:b9fn18m) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/25854cc3ed2748f2716a207962d2b9bb7476b50b/src/BloomExe/web/controllers/EditingViewApi.cs#L106)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

This is an odd comment on a line of code that makes it empty. Did you mean "can't be null"?  Oh, I see, you mean that the value passed to PostJson can't be empty, but that's an option we need here so "<NONE>" is used to mean it. Maybe on the previous line "a substitute for empty string, which is not allowed to be the result of RequiredPostJson"
</blockquote></details>

moved the comment and changed it.  A RequiredPostString cannot be empty.

**Review by JohnThomson — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289)* status: all files reviewed, 2 unresolved discussions (waiting on @andrew-polk)

___
*[src/BloomExe/web/controllers/EditingViewApi.cs, line 91 at r1](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289#-MW13Rhl6ChZacaU6eZC:-MW53hnr98MnHpE7kzFh:b-jx0nht) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/25854cc3ed2748f2716a207962d2b9bb7476b50b/src/BloomExe/web/controllers/EditingViewApi.cs#L91)):*
<details><summary><i>Previously, StephenMcConnel (Steve McConnel) wrote…</i></summary><blockquote>

Done.  I changed to using an EventHandler in Editing Model.   (Autofac may be wonderful and perfect, but I don't understand how it works at all.)
</blockquote></details>

That's fine, but it seems out of place for EditingViewApi to know about a specific thing that the model needs to do in response to the event. It would be more natural for the model to subscribe to its own event, or (more briefly) to put this line in EditingModel.HandleEditPagePaintedEvent.

**Review by andrew-polk — COMMENTED** [review]
> Reviewed 3 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @StephenMcConnel)

**Review by andrew-polk — COMMENTED** [review]
> Reviewed 1 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @StephenMcConnel)

**Review by StephenMcConnel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @StephenMcConnel)

___
*[src/BloomExe/web/controllers/EditingViewApi.cs, line 91 at r1](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289#-MW13Rhl6ChZacaU6eZC:-MW55FyQ8g-WuWgtSI7J:bl792ce) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/25854cc3ed2748f2716a207962d2b9bb7476b50b/src/BloomExe/web/controllers/EditingViewApi.cs#L91)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

That's fine, but it seems out of place for EditingViewApi to know about a specific thing that the model needs to do in response to the event. It would be more natural for the model to subscribe to its own event, or (more briefly) to put this line in EditingModel.HandleEditPagePaintedEvent.
</blockquote></details>

moved line to EditingModel.HandleEditPagePaintedEvent.

**Review by JohnThomson — APPROVED** [review]
> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @StephenMcConnel)

**Review by andrew-polk — APPROVED** [review]
> Reviewed 1 of 2 files at r3, 4 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4289)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @StephenMcConnel)
